Add experimental feature for post-processing Canvas image data

sysrqb opened the pull request:
#### 1a169d095a6691b2461810e3b639fffed1cf87f6
<pre>
Add experimental feature for post-processing Canvas image data
<a href="https://bugs.webkit.org/show_bug.cgi?id=243555">https://bugs.webkit.org/show_bug.cgi?id=243555</a>
rdar://98525861

Reviewed by Wenson Hsieh.

In this new experimental feature we post-process the Canvas2D and WebGL buffers
by adding small amounts of noise in the rendered image data. The approach this
patch takes is similar to a proposed patch by Eli Lucherini. In that patch, for
every pixel of the rendered canvas, with some probability one primary color
channel of a pixel might have a small amount of noise added. In this patch,
instead we add small amounts of noise depending on the whole pixel&apos;s value.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::makeRenderingResultsAvailable):
(WebCore::CanvasBase::postProcessPixelBuffer const):
* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getImageData):
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::postProcessPixelBuffer const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::setFillStyle):
(WebCore::computeImageDataRect):
(WebCore::CanvasRenderingContext2DBase::postProcessPixelBuffer const):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::suppliedColors const):
* Source/WebCore/html/canvas/CanvasStyle.h:
(WebCore::CanvasStyle::srgbaColor const):

Canonical link: <a href="https://commits.webkit.org/261197@main">https://commits.webkit.org/261197@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bab07e3bd5a951168deccaa9a224ce2d3f908e8e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110848 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19934 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/2199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114809 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/21359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/11071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/2199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116599 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/21359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/103215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/21359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/103215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/99480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12529 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/103215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/10627 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/11071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100581 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/18487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31370 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/7778 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/15031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108611 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4238 "Built successfully and passed tests") | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26773 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -1351,6 +1351,20 @@ CanvasFiltersEnabled:
     WebCore:
       default: false
 
+CanvasNoiseInjectionEnabled:
```
Is this separate feature flag needed? (Since this is already controlled by the same feature flag, which guards the presence of the hash salt).

sysrqb replied:
Adding a dedicated flag for this was a recommendation I got, given the current uncertainty around it.

whsieh replied:
Okay, sounds good to me.

whsieh commented on Source/WebCore/html/CanvasBase.cpp:
```diff
@@ -356,6 +358,88 @@ RefPtr<ImageBuffer> CanvasBase::allocateImageBuffer(bool usesDisplayListDrawing,
+
+        auto pixelHash = computeHash(salt, redChannel, greenChannel, blueChannel, alphaChannel);
+
+        auto clampedFivePercent = static_cast<int>((static_cast<float>(pixelHash / std::numeric_limits<uint32_t>::max()))*26 - 13);
```
Nit - spaces around the `*`.

```suggestion
        auto clampedFivePercent = static_cast<int>((static_cast<float>(pixelHash / std::numeric_limits<uint32_t>::max())) * 26 - 13);
```

whsieh commented on Source/WebCore/html/CanvasBase.cpp:
```diff
@@ -356,6 +358,88 @@ RefPtr<ImageBuffer> CanvasBase::allocateImageBuffer(bool usesDisplayListDrawing,
+        auto& redChannel = bytes[i];
+        auto& greenChannel = bytes[i+1];
+        auto& blueChannel = bytes[i+2];
+        auto& alphaChannel = bytes[i+3];
```
Is the format of this `PixelBuffer` always going to be RGBA? (If so, we could perhaps add an `ASSERT(pixelBuffer->format().pixelFormat == PixelFormat::RGBA8);`)

sysrqb replied:
I believe it will be, and we'll end up with garbage data if it's not but we handle it as if it is below...so we should add the assert.

whsieh commented on Source/WebCore/html/CanvasBase.cpp:
```diff
@@ -356,6 +358,88 @@ RefPtr<ImageBuffer> CanvasBase::allocateImageBuffer(bool usesDisplayListDrawing,
+            if (clampedFivePercent + alphaChannel > std::numeric_limits<uint8_t>::max())
+                clampedFivePercent = std::numeric_limits<uint8_t>::max() - alphaChannel;
+            else if (clampedFivePercent + alphaChannel < 0)
+                clampedFivePercent = -alphaChannel;
```
These 4 lines are repeated several times in this method; maybe we could add a helper lambda like:

```
auto clampedColorComponentOffset = [](int colorComponentOffset, int originalComponentValue) {
    if (colorComponentOffset + originalComponentValue > std::numeric_limits<uint8_t>::max())
        return std::numeric_limits<uint8_t>::max() - originalComponentValue;
    if (colorComponentOffset + originalComponentValue < 0)
        return -originalComponentValue;
    return colorComponentOffset;
};
```

...and then deploy it in these 5 places?

whsieh reviewed (approved):
Would be ideal to get some graphics experts to take a look as well — but we can start here.

sysrqb committed:
Add experimental feature for post-processing Canvas image data
https://bugs.webkit.org/show_bug.cgi?id=243555
rdar://98525861

Reviewed by Wenson Hsieh.

In this new experimental feature we post-process the Canvas2D and WebGL buffers
by adding small amounts of noise in the rendered image data. The approach this
patch takes is similar to a proposed patch by Eli Lucherini. In that patch, for
every pixel of the rendered canvas, with some probability one primary color
channel of a pixel might have a small amount of noise added. In this patch,
instead we add small amounts of noise depending on the whole pixel's value.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/CanvasBase.cpp:
(WebCore::CanvasBase::makeRenderingResultsAvailable):
(WebCore::CanvasBase::postProcessPixelBuffer const):
* Source/WebCore/html/CanvasBase.h:
* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::getImageData):
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::postProcessPixelBuffer const):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::setFillStyle):
(WebCore::computeImageDataRect):
(WebCore::CanvasRenderingContext2DBase::postProcessPixelBuffer const):
(WebCore::CanvasRenderingContext2DBase::didDraw):
(WebCore::CanvasRenderingContext2DBase::getImageData const):
(WebCore::CanvasRenderingContext2DBase::putImageData):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::suppliedColors const):
* Source/WebCore/html/canvas/CanvasStyle.h:
(WebCore::CanvasStyle::srgbaColor const):

Canonical link: https://commits.webkit.org/261197@main
